### DISCUSSION TITLE
Send data by QUIC server in arbitrary moment over existing connection

### DISCUSSION THREAD

**Discussion by jurcw** [aiortc/aioquic#337] [Q&A] [answered]
I'm trying to write a simple client + server applications, in which QUIC would act as a transport for other protocol on top of it (not HTTP/3).
The protocol I'm trying to send via QUIC is using it's own re-try mechanism, so I was planning to exchange data between QUIC client and server with the use of DATAGRAMs (`send_datagram_frame()`)

I took a look at the examples (e.q. DOQ client + server) and I also read the documentation and I've identified a way to send data from QUIC server only as a reaction to some other QUIC event (with the use of `quic_event_received()`. The nature of my protocol however requires QUIC server to send data towards client at any arbitrary moment (of course after the connection is established by the client's initiative). Is it something possible to be achieved with the use of aioquic library? If yes - I would really appreciate some guidance.

Thanks

**Comment by devkdas** [comment]
hai @exe32 

Could you please share your QUIC server and QUIC Client files where you are able to send data from QUIC Server to QUIC Client.

Thank you.
@Karthikdasari0423

**Comment by Skio22** [comment]
Hello, I am currently dealing with a similar problem as @exe32 so I would really appreciate some help here with this question.

**Reply by devkdas** [reply]
Could you please share your QUIC server and QUIC Client files where you are able to send data from QUIC Server to QUIC Client.

**Comment by jurcw** [comment]
@Karthikdasari0423 full program is not yet in shareable form, I'm only doing some experiments which are very similar to what you can find in doq_server.py example. Basically QUIC Server sends data back as a reaction to QUIC event.

Please find the cope snippet below:
```
    def quic_event_received(self, event: QuicEvent):
        if isinstance(event, DatagramFrameReceived):
            # parse query
            length = struct.unpack("!H", bytes(event.data[:2]))[0]
            query = event.data[2 : 2 + length].decode()
            print("received text: ", query)
            # send ACK back
            data = ("ACK" + " " + query).encode()   
            data = struct.pack("!H", len(data)) + data
            self._quic.send_datagram_frame(data)
```

I'm looking for an option here to send data from QUIC Server on demand at any point in time (once the connection is established of course).

**Comment by devkdas** [comment]
@exe32 
Hope this might helps

https://github.com/aiortc/aioquic/discussions/306

https://github.com/aiortc/aioquic/discussions/315

**Comment by Skio22** [comment, marked as answer]
I tried to implement my own QUIC server and client and I managed to figure it out, if you want to do it my way, you will need to use a little bit of threading. If you want to use this as a transport to another protocol like me, for example FTP or SSH server. Just initialize the QUIC connection as always, you can find it in the examples. On first quic_event_received function call, insert the identifier of event into some global Queue (identifier is in event.stream_id) variable and start a new thread (pass the receive QUIC data into the thread as argument). In the thread create a TCP socket for your TCP protocol, and send these data through TCP. After that, save the TCP socket connection to global variable, you will need that reference in the next quic_event_received. Finally make an infinite while loop in the thread, which will receive the data from socket with socket.recv() function. When the socket.recv() receive the TCP data back, check the Queue with stream ids. If the queue is not empty, then you know, that the QUIC client is waiting for some data. Therefore use the send_stream_data function with the identifier from queue, and received data from TCP to send data back to QUIC client through QUIC. In case, that the Queue is empty, you know, that client is not waiting for response, but your TCP protocol send you multiple responses to one request. In this case, call get_next_available_stream_id function, to generate the stream_id and send the data with new stream_id and received data back to client with send_stream_data function.

As I mentioned, you will need the socket connection reference. Use that reference in next quic_event_received calls to send received QUIC data to the TCP protocol

I assume that you are solving the similar problem as me. If you are not, this walkthrough is irrelevant and I am sorry. If you need some help you can contact me, and I can try to help you.

**Reply by devkdas** [reply]
hai @Skio22 

Thank you for replying back.

I am trying to solve the similar problem and looking for some help.

Could you please provide me your mail address or Could you please ping me on this mail address kartheekdasari1998@gmail.com so that we can have a discussion over there.

Looking forward for your reply.

Thank you.
@Karthikdasari0423

**Reply by jurcw** [reply]
Thank you @Skio22, your guidance helped me to achieve my goals. In deed - the key was to use threading and storing the instance of my protocol class somewhere globally (to easily access it in the other thread).

**Reply by devkdas** [reply]
Hai @exe32 

Would it be possible for  you to share your code to [kartheekdasari1998@gmail.com](mailto:kartheekdasari1998@gmail.com)

Thank you
@Karthikdasari0423
